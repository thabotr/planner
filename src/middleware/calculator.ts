const firstElevenPseudoFibonacciNumbers = [0, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89];
function toSubjectiveEffortScore(percent?: number): number {
    const zeroToTen = (percent ?? 0)/10;
    const intergralIndex = Math.floor(zeroToTen);
    return firstElevenPseudoFibonacciNumbers[intergralIndex];
}

export { toSubjectiveEffortScore };