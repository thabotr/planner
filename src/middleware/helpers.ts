
function percentToIntegralIndex(percent?: number): number {
    const zeroToTen = (percent ?? 0) / 10;
    return Math.floor(zeroToTen);
}
const firstElevenPseudoFibonacciNumbers = [0, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89];
function toSubjectiveEffortScore(percent?: number): number {
    const intergralIndex = percentToIntegralIndex(percent);
    return firstElevenPseudoFibonacciNumbers[intergralIndex];
}

function verboseTimestamp(diffInMs: number): string {
    const diffSecs = Math.floor(diffInMs / 1_000);
    const diffMins = Math.floor(diffSecs / 60);
    const diffHrs = Math.floor(diffMins / 60);
    const diffDays = Math.floor(diffHrs / 24);
    if (diffDays > 0) {
        return `${diffDays}dy`;
    }
    if (diffHrs > 0) {
        return `${diffHrs}h`;
    }
    if (diffMins > 0) {
        return `${diffMins}m`;
    }
    return `${diffSecs}s`;
}

const getRandomTime = () => Math.round(Math.random() * 1_000 * 60 * 60 * 24);

type TaskType = {
    pES: number,    // mental effort score
    mES: number,    // mental effort score
    length: number,     // time investment in milliseconds
}

type AvailabilityType = TaskType & {
    from: number,       // start time in milliseconds
};

type ScheduleItemType = {
    timeslot: AvailabilityType & { id: number },
    tasks: TaskType[],
};

class Scheduler {
    add(tslot: AvailabilityType): number | null {
        const item: ScheduleItemType = {
            timeslot: { ...tslot, id: this.#getId() },
            tasks: [],
        };

        const listIsEmpty = this.#list.length === 0;
        if (listIsEmpty) {
            this.#list.push(item);
            return item.timeslot.id;
        }

        let left = 0;
        let right = this.#list.length - 1;

        let middle: number;
        while (right - left > 1) {
            middle = Math.floor((right - left) / 2);
            if (this.#overlapsWithItemAt(tslot, left)) { return null; }
            if (this.#goesBeforeItemAt(tslot, middle)) { right = middle; }
            else { left = middle; }
        }

        if (this.#overlapsWithItemAt(tslot, left)) { return null; }

        if (this.#goesBeforeItemAt(tslot, left)) { this.#list.unshift(item); }
        else { this.#list.push(item); }

        return item.timeslot.id;
    }
    schedule(task: TaskType): number | null {
        for (let tslotIndex = 0; tslotIndex < this.#list.length; ++tslotIndex) {
            const scheduleTasks = this.#list[tslotIndex].tasks;
            const allTasks = scheduleTasks.concat(task);
            const superTask = allTasks.reduce((prev, curr) => ({
                length: prev.length + curr.length,
                mES: prev.mES + curr.mES,
                pES: prev.pES + curr.pES,
            }));
            const timeslot = this.#list[tslotIndex].timeslot;
            const superTaskFitsIntoTimeslot = timeslot.length >= superTask.length &&
                timeslot.mES >= superTask.mES && timeslot.pES >= superTask.pES

            if (superTaskFitsIntoTimeslot) {
                this.#list[tslotIndex].tasks = allTasks;
                return tslotIndex;
            }

        }
        return null;
    }
    getScheduleOn(day: Date): ScheduleItemType[] {
        const dayMask: AvailabilityType = {
            from: day.getTime(),
            length: AvailabilityDS.ADayInMillis,
            mES: 0,
            pES: 0,
        };
        return this.#list.filter((_, index) => this.#overlapsWithItemAt(dayMask, index));
    }
    remove(timeslotId: number): TaskType[] | null {
        let foundTslot2Rmv = false;
        for (const index in this.#list) {
            const { timeslot } = this.#list[index];
            foundTslot2Rmv = timeslot.id === timeslotId;
            if (foundTslot2Rmv) {
                const indexOfTSlot2Rmv = Number.parseInt(index);
                const lastItemIndex = this.#list.length - 1;
                const tasks4Tslot2rmv = this.#list[indexOfTSlot2Rmv].tasks;
                this.#list[indexOfTSlot2Rmv] = this.#list[lastItemIndex];
                this.#list.pop();
                return tasks4Tslot2rmv;
            }
        }
        return null;
    }
    #goesBeforeItemAt(tslot: AvailabilityType, index: number): boolean {
        const { timeslot: tslotAtIndex } = this.#list[index];
        return tslot.from < tslotAtIndex.from;
    }
    #overlapsWithItemAt(newTslot: AvailabilityType, index: number): boolean {
        const { timeslot } = this.#list[index];
        const newTslotStartContainedInTslot = timeslot.from <= newTslot.from &&
            newTslot.from <= timeslot.from + timeslot.length;
        const newTslotEndContainedInTslot = timeslot.from <= newTslot.from + newTslot.length &&
            newTslot.from + newTslot.length <= timeslot.from + timeslot.length;
        const newTslotContainsTslot = newTslot.from <= timeslot.from &&
            timeslot.from <= newTslot.from + newTslot.length;
        return newTslotStartContainedInTslot || newTslotEndContainedInTslot ||
            newTslotContainsTslot;
    }
    #id = 0;
    #getId() { ++this.#id; return this.#id - 1; }
    #list: ScheduleItemType[] = [];
}

class AvailabilityDS {
    // requirements
    // can add availability
    // cannot add availability if it intersects with already existing availability
    // should be able to provide us with this conflicting availability

    // can remove availability - should return it along with its scheduled contents else null

    // can schedule contents over some given availability provided the availability has enough capacity
    // can get scheduled contents and what part of the availability they consume

    // can split availability

    // can merge two adjacent availabilities else return false

    // can get availability given a single day
    list: Array<AvailabilityType> = [];
    #advancedList: Array<{ availability: AvailabilityType, tasks: TaskType[] }> = [];

    addAvailability(item: AvailabilityType): AvailabilityType | null {
        const availabilityOverlapsWithItem = (av: AvailabilityType) => item.from >= av.from && item.from <= av.from + av.length;
        if (this.list.some(availabilityOverlapsWithItem)) {
            return null;
        }
        this.list.push(item);
        return item;
    }
    getAvailabilityForDay(day: Date): Array<AvailabilityType> {
        const dayStartTime = day.getTime();
        const dayEndTime = dayStartTime + AvailabilityDS.ADayInMillis;
        return this.list.filter(av => av.from >= dayStartTime && av.from < dayEndTime);
    }

    scheduleTask(task: TaskType): AvailabilityType | null {
        return this.list.find(av => av.length >= task.length && av.mES >= task.mES && av.pES >= task.pES) ?? null;
    }

    clearAllAvailability() {
        this.list = [];
    }

    static ADayInMillis = 1_000 * 60 * 60 * 24;
    static AnHourInMillis = 1_000 * 60 * 60;
    static AMinuteInMillis = 1_000 * 60;
}

export { toSubjectiveEffortScore, verboseTimestamp, getRandomTime, AvailabilityDS, Scheduler, type AvailabilityType, type TaskType };