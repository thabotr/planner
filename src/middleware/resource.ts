import type DescriptiveItemType from '../types/DescriptiveItemType';
class Resource {
    async getTasks(): Promise<Array<DescriptiveItemType>> {
        // make get request
        return [];
    }

    async createTask(task: DescriptiveItemType) {
        // make post request
    }
}

export {Resource};