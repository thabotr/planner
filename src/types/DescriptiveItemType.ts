import ItemType from "./ItemType";

class DescriptiveItemType extends ItemType {
    description: string;
    constructor(id: string, length: number, pES: number, mES: number, description: string) {
        super(length, mES, pES, id);
        this.description = description;
    }
}

export default DescriptiveItemType;