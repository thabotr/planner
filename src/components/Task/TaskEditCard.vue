<template>
    <GenericItemEditCardVue :item="task" item-type="TASK" @save="onSave">
        <v-form ref="form" id="description">
            <v-textarea v-model="tempDescription" aria-label="description" :rules="[descriptionGEQ5Chars]"
                label="What to do?" density="compact" clearable />
        </v-form>
    </GenericItemEditCardVue>
</template>

<script lang="ts">
import DescriptiveItemType from '@/types/DescriptiveItemType';
import type ItemType from '@/types/ItemType';
import GenericItemEditCardVue from './GenericItemEditCard.vue';

export default {
    props: {
        task: {
            type: DescriptiveItemType,
            required: true,
        },
    },
    data() {
        return {
            tempDescription: this.task.description,
        };
    },
    methods: {
        descriptionGEQ5Chars(value: string): true | string {
            if ((value || '').length >= 5) {
                return true;
            }
            return "description must be 5 or more characters";
        },
        async onSave(item: ItemType) {
            const { valid } = await (this.$refs.form as any).validate();
            if (valid) {
                const updatedTask: DescriptiveItemType = new DescriptiveItemType(
                    item.id, item.length, item.pES, item.mES, this.tempDescription
                );
                this.$emit("save", updatedTask);
            }
        }
    },
    components: { GenericItemEditCardVue, },
    emits: ['save'],
}
</script>

<style scoped>
#description {
    padding: 0.5rem;
}
</style>