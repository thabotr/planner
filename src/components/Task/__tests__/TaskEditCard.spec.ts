import TaskEditCard from "@/components/Task/TaskEditCard.vue";
import { TimeInMillis } from "@/middleware/helpers";
import DescriptiveItemType from "@/types/DescriptiveItemType";
import { describe, expect, it } from "vitest";
import { CustomVueWrapper } from "../../__tests__/utils";

describe('TaskEditCard', () => {
    const task = new DescriptiveItemType(
        "id", TimeInMillis.Hour, 10, 10, "description here"
    );
    const wrapper = new CustomVueWrapper(TaskEditCard, {
        props: {
            task: task,
        },
        slots: {
            default: 'Default',
            actions: 'Actions go here',
            controls: 'Card effort controls go here',
        },
    });
    it('renders correctly', () => {
        expect(wrapper.wrapper.html()).toMatchSnapshot();
    });
    it("emits 'save' with the updated task on click save", async () => {
        const updatedTask = new DescriptiveItemType(
            task.id, task.length, task.pES, task.mES, 'a different description',
        );
        expect(wrapper.wrapper.emitted()).not.toHaveProperty('save');
        await wrapper.findByLabel('description').setValue(updatedTask.description);
        await wrapper.click('save');
        expect(wrapper.wrapper.emitted()).toHaveProperty('save');
        const [event] = wrapper.wrapper.emitted()['save'];
        const [emittedTask] = event as any[];
        expect(emittedTask).toStrictEqual(updatedTask);
    });
})