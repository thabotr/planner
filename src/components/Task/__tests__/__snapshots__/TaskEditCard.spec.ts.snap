// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TaskEditCard > renders correctly 1`] = `
"<div data-v-f39f1573=\\"\\" data-v-9b7145d6=\\"\\" data-v-bf677fdb=\\"\\">
  <div data-v-f39f1573=\\"\\" class=\\"v-card v-theme--light v-card--density-default v-card--variant-elevated\\" id=\\"bg\\">
    <!---->
    <div class=\\"v-card__loader\\">
      <div class=\\"v-progress-linear v-theme--light\\" style=\\"top: 0px; height: 0px; --v-progress-linear-height: 2px; left: 50%; transform: translateX(-50%);\\" role=\\"progressbar\\" aria-hidden=\\"true\\" aria-valuemin=\\"0\\" aria-valuemax=\\"100\\">
        <!---->
        <div class=\\"v-progress-linear__background\\" style=\\"width: 100%;\\"></div>
        <transition-stub name=\\"fade-transition\\" appear=\\"false\\" persisted=\\"false\\" css=\\"true\\">
          <div class=\\"v-progress-linear__indeterminate\\">
            <div class=\\"v-progress-linear__indeterminate long\\"></div>
            <div class=\\"v-progress-linear__indeterminate short\\"></div>
          </div>
        </transition-stub>
        <!---->
      </div>
    </div>
    <!---->
    <!---->
    <div data-v-f39f1573=\\"\\" class=\\"v-card v-theme--light v-card--density-default v-card--variant-elevated\\" id=\\"fg\\">
      <!---->
      <div class=\\"v-card__loader\\">
        <div class=\\"v-progress-linear v-theme--light\\" style=\\"top: 0px; height: 0px; --v-progress-linear-height: 2px; left: 50%; transform: translateX(-50%);\\" role=\\"progressbar\\" aria-hidden=\\"true\\" aria-valuemin=\\"0\\" aria-valuemax=\\"100\\">
          <!---->
          <div class=\\"v-progress-linear__background\\" style=\\"width: 100%;\\"></div>
          <transition-stub name=\\"fade-transition\\" appear=\\"false\\" persisted=\\"false\\" css=\\"true\\">
            <div class=\\"v-progress-linear__indeterminate\\">
              <div class=\\"v-progress-linear__indeterminate long\\"></div>
              <div class=\\"v-progress-linear__indeterminate short\\"></div>
            </div>
          </transition-stub>
          <!---->
        </div>
      </div>
      <!---->
      <!---->
      <div data-v-9b7145d6=\\"\\" class=\\"flex-horizontal gapped-s\\"><span data-v-9b7145d6=\\"\\" class=\\"v-chip v-theme--light v-chip--density-compact v-chip--size-default v-chip--variant-tonal\\" draggable=\\"false\\"><!----><span class=\\"v-chip__underlay\\"></span>
        <!---->
        <!---->TASK
        <!---->
        <!----></span>
      </div>
      <form data-v-bf677fdb=\\"\\" class=\\"v-form\\" novalidate=\\"\\" id=\\"description\\">
        <div data-v-bf677fdb=\\"\\" class=\\"v-input v-input--horizontal v-input--density-compact v-text-field\\">
          <!---->
          <div class=\\"v-input__control\\">
            <div class=\\"v-field v-field--appended v-field--variant-filled v-theme--light\\" role=\\"textbox\\">
              <div class=\\"v-field__overlay\\"></div>
              <div class=\\"v-field__loader\\">
                <div class=\\"v-progress-linear v-theme--light\\" style=\\"top: 0px; height: 0px; --v-progress-linear-height: 2px; left: 50%; transform: translateX(-50%);\\" role=\\"progressbar\\" aria-hidden=\\"true\\" aria-valuemin=\\"0\\" aria-valuemax=\\"100\\">
                  <!---->
                  <div class=\\"v-progress-linear__background\\" style=\\"width: 100%;\\"></div>
                  <transition-stub name=\\"fade-transition\\" appear=\\"false\\" persisted=\\"false\\" css=\\"true\\">
                    <div class=\\"v-progress-linear__indeterminate\\">
                      <div class=\\"v-progress-linear__indeterminate long\\"></div>
                      <div class=\\"v-progress-linear__indeterminate short\\"></div>
                    </div>
                  </transition-stub>
                  <!---->
                </div>
              </div>
              <!---->
              <div class=\\"v-field__field\\" data-no-activator=\\"\\"><label class=\\"v-label v-field-label v-field-label--floating\\" aria-hidden=\\"true\\" for=\\"input-1\\">
                  <!---->What to do?
                </label><label class=\\"v-label v-field-label\\" for=\\"input-1\\">
                  <!---->What to do?
                </label>
                <!----><input size=\\"1\\" type=\\"text\\" id=\\"input-1\\" aria-describedby=\\"input-1-messages\\" aria-label=\\"description\\" class=\\"v-field__input\\">
                <!---->
              </div>
              <transition-stub name=\\"expand-x-transition\\" appear=\\"false\\" persisted=\\"false\\" css=\\"true\\">
                <div class=\\"v-field__clearable\\" style=\\"display: none;\\"><i class=\\"mdi-close-circle mdi v-icon notranslate v-theme--light v-icon--size-default v-icon--clickable\\" role=\\"button\\" aria-hidden=\\"false\\" aria-label=\\"Clear What to do?\\"></i></div>
              </transition-stub>
              <div class=\\"v-field__append-inner\\">
                <!---->
                <!---->
              </div>
              <div class=\\"v-field__outline\\">
                <!---->
                <!---->
              </div>
            </div>
          </div>
          <!---->
          <div class=\\"v-input__details\\">
            <transition-group-stub name=\\"slide-y-transition\\" tag=\\"div\\" appear=\\"false\\" persisted=\\"false\\" css=\\"true\\" class=\\"v-messages\\" role=\\"alert\\" aria-live=\\"polite\\" id=\\"input-1-messages\\">
              <!---->
            </transition-group-stub>
            <!---->
          </div>
        </div>
      </form>
      <div data-v-f39f1573=\\"\\" class=\\"flex-horizontal\\">
        <div data-v-19720537=\\"\\" data-v-f39f1573=\\"\\">
          <div data-v-19720537=\\"\\" class=\\"v-list-item v-theme--light v-list-item--density-default v-list-item--variant-text\\">
            <!----><span class=\\"v-list-item__underlay\\"></span>
            <div class=\\"v-list-item__prepend\\">
              <!---->
              <!----><i data-v-19720537=\\"\\" class=\\"mdi-brain mdi v-icon notranslate v-theme--light v-icon--size-default\\" aria-hidden=\\"true\\" id=\\"i-mental\\"></i>
            </div>
            <div class=\\"v-list-item__content\\" data-no-activator=\\"\\">
              <!---->
              <!---->
              <div data-v-19720537=\\"\\" class=\\"v-list-item-title\\">1 mental EP</div>
            </div>
            <!---->
          </div>
          <div data-v-19720537=\\"\\" class=\\"v-list-item v-theme--light v-list-item--density-default v-list-item--variant-text\\">
            <!----><span class=\\"v-list-item__underlay\\"></span>
            <div class=\\"v-list-item__prepend\\">
              <!---->
              <!----><i data-v-19720537=\\"\\" class=\\"mdi-account-hard-hat mdi v-icon notranslate v-theme--light v-icon--size-default\\" aria-hidden=\\"true\\" id=\\"i-physical\\"></i>
            </div>
            <div class=\\"v-list-item__content\\" data-no-activator=\\"\\">
              <!---->
              <!---->
              <div data-v-19720537=\\"\\" class=\\"v-list-item-title\\">1 physical EP</div>
            </div>
            <!---->
          </div>
          <div data-v-19720537=\\"\\" class=\\"v-list-item v-theme--light v-list-item--density-default v-list-item--variant-text\\">
            <!----><span class=\\"v-list-item__underlay\\"></span>
            <div class=\\"v-list-item__prepend\\">
              <!---->
              <!----><i data-v-19720537=\\"\\" class=\\"mdi-timer mdi v-icon notranslate v-theme--light v-icon--size-default\\" aria-hidden=\\"true\\" id=\\"i-time\\"></i>
            </div>
            <div class=\\"v-list-item__content\\" data-no-activator=\\"\\">
              <!---->
              <!---->
              <div data-v-19720537=\\"\\" class=\\"v-list-item-title\\">1h</div>
            </div>
            <!---->
          </div>
        </div>
        <div data-v-9b7145d6=\\"\\" id=\\"controls\\" class=\\"flex-vertical\\">
          <div data-v-9b7145d6=\\"\\" id=\\"mes-control\\"><button data-v-9b7145d6=\\"\\" type=\\"button\\" class=\\"v-btn v-btn--flat v-btn--icon v-theme--light v-btn--density-compact v-btn--size-default v-btn--variant-elevated\\" aria-label=\\"decrement mes\\"><span class=\\"v-btn__overlay\\"></span><span class=\\"v-btn__underlay\\"></span>
              <!----><span class=\\"v-btn__content\\" data-no-activator=\\"\\"><i class=\\"mdi-minus mdi v-icon notranslate v-theme--light v-icon--size-default\\" aria-hidden=\\"true\\"></i></span>
              <!---->
              <!---->
            </button><button data-v-9b7145d6=\\"\\" type=\\"button\\" class=\\"v-btn v-btn--flat v-btn--icon v-theme--light v-btn--density-compact v-btn--size-default v-btn--variant-elevated\\" aria-label=\\"increment mes\\"><span class=\\"v-btn__overlay\\"></span><span class=\\"v-btn__underlay\\"></span>
              <!----><span class=\\"v-btn__content\\" data-no-activator=\\"\\"><i class=\\"mdi-plus mdi v-icon notranslate v-theme--light v-icon--size-default\\" aria-hidden=\\"true\\"></i></span>
              <!---->
              <!---->
            </button></div>
          <div data-v-9b7145d6=\\"\\" id=\\"pes-control\\"><button data-v-9b7145d6=\\"\\" type=\\"button\\" class=\\"v-btn v-btn--flat v-btn--icon v-theme--light v-btn--density-compact v-btn--size-default v-btn--variant-elevated\\" aria-label=\\"decrement pes\\"><span class=\\"v-btn__overlay\\"></span><span class=\\"v-btn__underlay\\"></span>
              <!----><span class=\\"v-btn__content\\" data-no-activator=\\"\\"><i class=\\"mdi-minus mdi v-icon notranslate v-theme--light v-icon--size-default\\" aria-hidden=\\"true\\"></i></span>
              <!---->
              <!---->
            </button><button data-v-9b7145d6=\\"\\" type=\\"button\\" class=\\"v-btn v-btn--flat v-btn--icon v-theme--light v-btn--density-compact v-btn--size-default v-btn--variant-elevated\\" aria-label=\\"increment pes\\"><span class=\\"v-btn__overlay\\"></span><span class=\\"v-btn__underlay\\"></span>
              <!----><span class=\\"v-btn__content\\" data-no-activator=\\"\\"><i class=\\"mdi-plus mdi v-icon notranslate v-theme--light v-icon--size-default\\" aria-hidden=\\"true\\"></i></span>
              <!---->
              <!---->
            </button></div>
          <div data-v-9b7145d6=\\"\\" id=\\"length-control\\"><button data-v-9b7145d6=\\"\\" type=\\"button\\" class=\\"v-btn v-btn--flat v-btn--icon v-theme--light v-btn--density-compact v-btn--size-default v-btn--variant-elevated\\" aria-label=\\"decrement time\\"><span class=\\"v-btn__overlay\\"></span><span class=\\"v-btn__underlay\\"></span>
              <!----><span class=\\"v-btn__content\\" data-no-activator=\\"\\"><i class=\\"mdi-undo mdi v-icon notranslate v-theme--light v-icon--size-default\\" aria-hidden=\\"true\\"></i></span>
              <!---->
              <!---->
            </button><button data-v-9b7145d6=\\"\\" type=\\"button\\" class=\\"v-btn v-btn--flat v-btn--icon v-theme--light v-btn--density-compact v-btn--size-default v-btn--variant-elevated\\" aria-label=\\"increment time\\"><span class=\\"v-btn__overlay\\"></span><span class=\\"v-btn__underlay\\"></span>
              <!----><span class=\\"v-btn__content\\" data-no-activator=\\"\\"><i class=\\"mdi-fast-forward-5 mdi v-icon notranslate v-theme--light v-icon--size-default\\" aria-hidden=\\"true\\"></i></span>
              <!---->
              <!---->
            </button></div>
        </div>
      </div>
      <!---->
      <!----><span class=\\"v-card__underlay\\"></span>
    </div>
    <div data-v-f39f1573=\\"\\" id=\\"actions\\"><span data-v-f39f1573=\\"\\" class=\\"v-chip v-theme--light v-chip--density-compact v-chip--size-default v-chip--variant-tonal\\" draggable=\\"false\\"><!----><span class=\\"v-chip__underlay\\"></span>
      <!---->
      <!---->id
      <!---->
      <!----></span>
      <div data-v-9b7145d6=\\"\\" class=\\"flex-vertical justify-space-between all-actions\\"><button data-v-9b7145d6=\\"\\" type=\\"button\\" class=\\"v-btn v-btn--flat v-btn--icon v-theme--light v-btn--density-compact v-btn--size-default v-btn--variant-elevated\\" aria-label=\\"delete\\"><span class=\\"v-btn__overlay\\"></span><span class=\\"v-btn__underlay\\"></span>
          <!----><span class=\\"v-btn__content\\" data-no-activator=\\"\\"><i class=\\"mdi-delete-forever mdi v-icon notranslate v-theme--light v-icon--size-default\\" aria-hidden=\\"true\\"></i></span>
          <!---->
          <!---->
        </button>
        <div data-v-9b7145d6=\\"\\" class=\\"flex-vertical safe-actions\\"><button data-v-9b7145d6=\\"\\" type=\\"button\\" class=\\"v-btn v-btn--flat v-btn--icon v-theme--light v-btn--density-compact v-btn--size-default v-btn--variant-elevated\\" aria-label=\\"cancel\\"><span class=\\"v-btn__overlay\\"></span><span class=\\"v-btn__underlay\\"></span>
            <!----><span class=\\"v-btn__content\\" data-no-activator=\\"\\"><i class=\\"mdi-close mdi v-icon notranslate v-theme--light v-icon--size-default\\" aria-hidden=\\"true\\"></i></span>
            <!---->
            <!---->
          </button><button data-v-9b7145d6=\\"\\" type=\\"button\\" class=\\"v-btn v-btn--flat v-btn--icon v-theme--light v-btn--density-compact v-btn--size-default v-btn--variant-elevated\\" aria-label=\\"save\\"><span class=\\"v-btn__overlay\\"></span><span class=\\"v-btn__underlay\\"></span>
            <!----><span class=\\"v-btn__content\\" data-no-activator=\\"\\"><i class=\\"mdi-content-save mdi v-icon notranslate v-theme--light v-icon--size-default\\" aria-hidden=\\"true\\"></i></span>
            <!---->
            <!---->
          </button></div>
      </div>
    </div>
    <!---->
    <!----><span class=\\"v-card__underlay\\"></span>
  </div>
</div>"
`;
